src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,253 @@
+package htsjdk.samtools.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Implementation of an asynchronous writer pool.
+ *
+ * @author Seth Stadick
+ */
+public class AsyncWriterPool implements Closeable {
+    private final ExecutorService executor;
+    private final AtomicBoolean poolClosed = new AtomicBoolean(false);
+    private final List<PooledWriter<?>> writers = new ArrayList<>();
+
+    /**
+     * Create an AsyncWriterPool using the specified number of {@code threads}. The number of {@code threads} in use at
+     * one time will grow and shrink with load.
+     *
+     * @param threads max number of threads to use
+     */
+    public AsyncWriterPool(final int threads) {
+        if (threads < 1) throw new IllegalArgumentException("Threads must be >= 1: " + threads);
+
+        this.executor = Executors.newWorkStealingPool(threads);
+    }
+
+    /**
+     * Create an AsyncWriterPool using all available processors. The number of threads in use at one time will grow and
+     * shrink with load.
+     */
+    public AsyncWriterPool() {
+        this(Runtime.getRuntime().availableProcessors());
+    }
+
+    /**
+     * Asynchronously closes each writer in the pool. Each writer calls {@code .close()} in a CompletableFuture, which
+     * allows the writers to wait for any ongoing writes, drain any remaining elements from the queue, and then close
+     * the inner writer. All writers will immediately cease to accept new items to write and any call to {@code .write()}
+     * after calling this method will throw an exception. This method waits for all {@code CompletableFuture}s to
+     * complete and then shuts down the executor.
+     * <p>
+     * This method blocks and till all writers have closed and pool has shut down.
+     *
+     * @throws IOException if any writer raises an exception
+     */
+    @Override
+    public void close() throws IOException {
+        this.poolClosed.set(true);
+        List<CompletableFuture<Void>> closers = new ArrayList<>();
+        for (final PooledWriter<?> writer : this.writers) {
+            closers.add(writer.asyncClose());
+        }
+        // Convert List to array and wait on all futures to complete
+        CompletableFuture.allOf(closers.toArray(new CompletableFuture[0])).join();
```

Comment:
Will this handle things gracefully if one or more writers have already been `closed()` directly?

Reply:
No, I think as it is here, there is a chance that `AsyncWriterPool.close` could launch the `nonBlockingClose` and then someone could call `close` on the writer leading to two threads modifying things and two threads writing at the same time. I've changed it so that now:

1. if isClosed is true, and either `nonBlockingClose` or `close` are called it will result in an exception. 
2. `nonBlockingClose` now immediatly sets `isClosed` to true so that two "closing" pathways can't be running at the same time.
3. I've followed the same pattern as `checkAndRethrow` so that there is now a `blocking` and `nonBlocking` close method.